examples/generative/wgan_gp.py

```diff
@@ -0,0 +1,430 @@
+"""
+Title: WGAN-GP overriding `Model.train_step`
+Author: [A_K_Nain](https://twitter.com/A_K_Nain)
+Date created: 2020/05/9
+Last modified: 2020/05/9
+Description: WGAN-GP implementation
+"""
+
+"""
+## Wasserstein GAN (WGAN) with Gradient Penalty (GP)
+
+The original [Wasserstein GAN](https://arxiv.org/abs/1701.07875) leverages
+the Wasserstein distance to produce a value function that has better theoretical
+properties than the original.  WGAN requires that the discriminator (aka the critic)
+must lie within the space of 1-Lipschitz functions. The authors proposed the idea of
+weight clipping to achieve the same. Though weight clipping works, it is not a very
+good idea to enforce 1-Lipschitz and can cause undesirable behavior.
+
+[WGAN-GP](https://arxiv.org/pdf/1704.00028.pdf) proposed an alternative to weight
+clipping to ensure smooth training. Instead of clipping the weights, the authors
+proposed `gradient penalty` Gradient penalty measures the squared difference
+between the norm of the gradient of the predictions and 1. 
+"""
+
+"""
+## Setup
+"""
+
+import numpy as np
+import tensorflow as tf
+from tensorflow import keras
+from tensorflow.keras import layers
+
+
+"""
+## Prepare Fashion-MNIST data
+
+We will be using the [Fashion-MNIST](https://github.com/zalandoresearch/fashion-mnist) dataset
+in this work to demonstrate the training of WGAN-GP. Each sample in this dataset is a 28x28
+grayscale image associated with a label from 10 classes. 
+"""
+
+IMG_SHAPE = (28, 28, 1)
+BUFFER_SIZE = 60000
+BATCH_SIZE = 512
+
+# Size of noise vector
+noise_dim = 128
+
+fashion_mnist = keras.datasets.fashion_mnist
+(train_images, train_labels), (test_images, test_labels) = fashion_mnist.load_data()
+print(f"Number of examples: {len(train_images)}")
+print(f"Shape of the images in the dataset: {train_images.shape[1:]}")
+
+# we will reshape each sample to (28, 28, 1) and normalize the pixel values in [-1, 1].
+train_images = train_images.reshape(train_images.shape[0], *IMG_SHAPE).astype("float32")
+train_images = (train_images - 127.5) / 127.5
+train_dataset = (
+    tf.data.Dataset.from_tensor_slices(train_images)
+    .shuffle(BUFFER_SIZE)
```

Comment:
If I directly use `train_images`, it throws errors. That's why I used `tf.data.Dataset`

Reply:
What's the error?

Reply:
```
Epoch 1/20
WARNING:tensorflow:Model was constructed with shape (None, 28, 28, 1) for input Tensor("input_1:0", shape=(None, 28, 28, 1), dtype=float32), but it was called on an input with incompatible shape (1, None, 28, 28).

---------------------------------------------------------------------------

ValueError                                Traceback (most recent call last)

<ipython-input-3-56fff80df7fb> in <module>()
    414 wgan.fit(train_images,
    415          batch_size=BATCH_SIZE,
--> 416          epochs=epochs, callbacks=[cbk])

10 frames

/usr/local/lib/python3.6/dist-packages/tensorflow/python/keras/engine/training.py in _method_wrapper(self, *args, **kwargs)
     64   def _method_wrapper(self, *args, **kwargs):
     65     if not self._in_multi_worker_mode():  # pylint: disable=protected-access
---> 66       return method(self, *args, **kwargs)
     67 
     68     # Running inside `run_distribute_coordinator` already.

/usr/local/lib/python3.6/dist-packages/tensorflow/python/keras/engine/training.py in fit(self, x, y, batch_size, epochs, verbose, callbacks, validation_split, validation_data, shuffle, class_weight, sample_weight, initial_epoch, steps_per_epoch, validation_steps, validation_batch_size, validation_freq, max_queue_size, workers, use_multiprocessing)
    846                 batch_size=batch_size):
    847               callbacks.on_train_batch_begin(step)
--> 848               tmp_logs = train_function(iterator)
    849               # Catch OutOfRangeError for Datasets of unknown size.
    850               # This blocks until the batch has finished executing.

/usr/local/lib/python3.6/dist-packages/tensorflow/python/eager/def_function.py in __call__(self, *args, **kwds)
    578         xla_context.Exit()
    579     else:
--> 580       result = self._call(*args, **kwds)
    581 
    582     if tracing_count == self._get_tracing_count():

/usr/local/lib/python3.6/dist-packages/tensorflow/python/eager/def_function.py in _call(self, *args, **kwds)
    625       # This is the first call of __call__, so we have to initialize.
    626       initializers = []
--> 627       self._initialize(args, kwds, add_initializers_to=initializers)
    628     finally:
    629       # At this point we know that the initialization is complete (or less

/usr/local/lib/python3.6/dist-packages/tensorflow/python/eager/def_function.py in _initialize(self, args, kwds, add_initializers_to)
    504     self._concrete_stateful_fn = (
    505         self._stateful_fn._get_concrete_function_internal_garbage_collected(  # pylint: disable=protected-access
--> 506             *args, **kwds))
    507 
    508     def invalid_creator_scope(*unused_args, **unused_kwds):

/usr/local/lib/python3.6/dist-packages/tensorflow/python/eager/function.py in _get_concrete_function_internal_garbage_collected(self, *args, **kwargs)
   2444       args, kwargs = None, None
   2445     with self._lock:
-> 2446       graph_function, _, _ = self._maybe_define_function(args, kwargs)
   2447     return graph_function
   2448 

/usr/local/lib/python3.6/dist-packages/tensorflow/python/eager/function.py in _maybe_define_function(self, args, kwargs)
   2775 
   2776       self._function_cache.missed.add(call_context_key)
-> 2777       graph_function = self._create_graph_function(args, kwargs)
   2778       self._function_cache.primary[cache_key] = graph_function
   2779       return graph_function, args, kwargs

/usr/local/lib/python3.6/dist-packages/tensorflow/python/eager/function.py in _create_graph_function(self, args, kwargs, override_flat_arg_shapes)
   2665             arg_names=arg_names,
   2666             override_flat_arg_shapes=override_flat_arg_shapes,
-> 2667             capture_by_value=self._capture_by_value),
   2668         self._function_attributes,
   2669         # Tell the ConcreteFunction to clean up its graph once it goes out of

/usr/local/lib/python3.6/dist-packages/tensorflow/python/framework/func_graph.py in func_graph_from_py_func(name, python_func, args, kwargs, signature, func_graph, autograph, autograph_options, add_control_dependencies, arg_names, op_return_value, collections, capture_by_value, override_flat_arg_shapes)
    979         _, original_func = tf_decorator.unwrap(python_func)
    980 
--> 981       func_outputs = python_func(*func_args, **func_kwargs)
    982 
    983       # invariant: `func_outputs` contains only Tensors, CompositeTensors,

/usr/local/lib/python3.6/dist-packages/tensorflow/python/eager/def_function.py in wrapped_fn(*args, **kwds)
    439         # __wrapped__ allows AutoGraph to swap in a converted function. We give
    440         # the function a weak reference to itself to avoid a reference cycle.
--> 441         return weak_wrapped_fn().__wrapped__(*args, **kwds)
    442     weak_wrapped_fn = weakref.ref(wrapped_fn)
    443 

/usr/local/lib/python3.6/dist-packages/tensorflow/python/framework/func_graph.py in wrapper(*args, **kwargs)
    966           except Exception as e:  # pylint:disable=broad-except
    967             if hasattr(e, "ag_error_metadata"):
--> 968               raise e.ag_error_metadata.to_exception(e)
    969             else:
    970               raise

ValueError: in user code:

    /usr/local/lib/python3.6/dist-packages/tensorflow/python/keras/engine/training.py:571 train_function  *
        outputs = self.distribute_strategy.run(
    <ipython-input-3-56fff80df7fb>:319 train_step  *
        gp = self.gradient_penalty(batch_size, real_images, fake_images)
    <ipython-input-3-56fff80df7fb>:277 gradient_penalty  *
        pred = self.discriminator(interpolated, training=True)
    /usr/local/lib/python3.6/dist-packages/tensorflow/python/keras/engine/base_layer.py:927 __call__  **
        outputs = call_fn(cast_inputs, *args, **kwargs)
    /usr/local/lib/python3.6/dist-packages/tensorflow/python/keras/engine/network.py:719 call
        convert_kwargs_to_constants=base_layer_utils.call_context().saving)
    /usr/local/lib/python3.6/dist-packages/tensorflow/python/keras/engine/network.py:888 _run_internal_graph
        output_tensors = layer(computed_tensors, **kwargs)
    /usr/local/lib/python3.6/dist-packages/tensorflow/python/keras/engine/base_layer.py:886 __call__
        self.name)
    /usr/local/lib/python3.6/dist-packages/tensorflow/python/keras/engine/input_spec.py:216 assert_input_compatibility
        ' but received input with shape ' + str(shape))

    ValueError: Input 0 of layer conv2d is incompatible with the layer: expected axis -1 of input shape to have value 1 but received input with shape [1, None, 32, 28]



```

Reply:
I see. You should be able to fix it by starting `train_step` with `real_images = real_images[0]` or perhaps `if isinstance(real_images, tuple): real_images = real_images[0]`.

Reply:
Yeah, my bad 🤦 . Anyways, without `preftech`, each epoch takes more than 3 minutes and we need to run at least 10 epochs to get reasonable results.

Reply:
Our bad. This is a gotcha that we intend on fixing.

Reply:
Anyways, I have fixed it.